rosetta/transactions/intent.go

```diff
@@ -0,0 +1,175 @@
+		}
+	}
+
+	// Assume the first operation is the one with the negative amount.
+	send := operations[0]
+	receive := operations[1]
+
+	// If that was not the case, switch the send and receive operations.
+	if !strings.HasPrefix(send.Amount.Value, "-") {
+		receive = operations[0]
+		send = operations[1]
+	}
+
+	// Validate the sender and the receiver account IDs.
+	err := p.validate.Account(send.AccountID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid sender account: %w", err)
+	}
+	err = p.validate.Account(receive.AccountID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid receiver account: %w", err)
+	}
+
+	// Validate the currencies specified for deposit and withdrawal.
+	send.Amount.Currency, err = p.validate.Currency(send.Amount.Currency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid sender currency: %w", err)
+	}
+	receive.Amount.Currency, err = p.validate.Currency(receive.Amount.Currency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid receiver currency: %w", err)
+	}
+
+	// Make sure that both the send and receive operations use the same currency.
+	// This is perhaps unnecessary at the moment since we only have a single currency.
+	if send.Amount.Currency != receive.Amount.Currency {
+		return nil, failure.InvalidIntent{
+			Sender:      send.AccountID.Address,
+			Receiver:    receive.AccountID.Address,
+			Description: failure.NewDescription("send and receive currencies do not match"),
+		}
+	}
+
+	// Parse value specified by the sender, after removing the negative sign prefix.
+	trimmed := strings.TrimPrefix(send.Amount.Value, "-")
+	sv, err := convert.ParseRosettaValue(trimmed)
+	if err != nil {
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("could not parse withdrawal amount",
+				failure.WithString("withdrawal_amount", send.Amount.Value),
+				failure.WithErr(err),
+			),
+		}
+	}
+
+	// Parse value specified by the receiver.
+	rv, err := convert.ParseRosettaValue(receive.Amount.Value)
+	if err != nil {
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("could not parse deposit amount",
+				failure.WithString("deposit_amount", receive.Amount.Value),
+				failure.WithErr(err),
+			),
+		}
+	}
+
+	// Check if the specified amounts match.
+	if sv != rv {
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("deposit and withdrawal amounts do not match",
+				failure.WithString("deposit_amount", receive.Amount.Value),
+				failure.WithString("withdrawal_amount", send.Amount.Value),
+			),
+		}
+	}
+
+	// Validate that the specified operations are transfers.
+	if strings.ToUpper(send.Type) != dps.OperationTransfer ||
+		strings.ToUpper(receive.Type) != dps.OperationTransfer {
```

Comment:
We shouldn't have to do the `strings.ToUpper` here, the input should already be correct.

Reply:
In the rosetta docs, these values are represented as `Transfer`, while our consts use `TRANSFER`. So I uppercased it in order for it to be case insensitive match..

Reply:
The Rosetta specification only gives examples. The types of operations are defined by the Rosetta API implementation. So whatever we define the operation types to be is what should be used by a consumer.